includes/models/model.llms.student.php

```diff
@@ -1433,38 +1434,55 @@ public function mark_incomplete( $object_id, $object_type, $trigger = 'unspecifi
 
-		// Remove the user from the membership level.
+		global $wpdb;
+
+		// Remove the user from the given membership level.
 		$membership_levels = $this->get_membership_levels();
-		$key               = array_search( $membership_id, $membership_levels );
+		$key               = array_search( $membership_id, $membership_levels, true );
 		if ( false !== $key ) {
 			unset( $membership_levels[ $key ] );
 		}
 		update_user_meta( $this->get_id(), '_llms_restricted_levels', $membership_levels );
 
-		global $wpdb;
-		// Locate all enrollments triggered by this membership level.
-		$q = $wpdb->get_results(
+		// Fetch all of this user's enrolled course IDs triggered by any membership.
+		$course_ids_enrolled_by_any_membership = $wpdb->get_col(
 			$wpdb->prepare(
-				"SELECT post_id FROM {$wpdb->prefix}lifterlms_user_postmeta WHERE user_id = %d AND meta_key = '_enrollment_trigger' AND meta_value = %s",
-				array( $this->get_id(), 'membership_' . $membership_id )
-			),
-			'OBJECT_K'
+				"SELECT post_id FROM {$wpdb->prefix}lifterlms_user_postmeta " .
+				"WHERE user_id = %d AND meta_key = '_enrollment_trigger' AND meta_value LIKE %s",
+				array( $this->get_id(), 'membership_%' )
+			)
 		); // db call ok; no-cache ok.
+		$course_ids_enrolled_by_any_membership = array_map( 'intval', $course_ids_enrolled_by_any_membership ); // Cast strings to integers.
 
-		$courses = array_keys( $q );
+		// Get the unique auto-enrollment course IDs, with any post status, of other membership levels.
+		$other_memberships_auto_enroll_course_ids = array();
+		foreach ( $membership_levels as $membership_level ) {
+			$membership_auto_enroll_course_ids         = ( new LLMS_Membership( $membership_level ) )->get( 'auto_enroll' );
+			$other_memberships_auto_enroll_course_ids += array_flip( $membership_auto_enroll_course_ids );
+		}
 
-		if ( $courses ) {
+		// Get the auto-enrollment course IDs, with any post status, of the given membership.
+		$given_membership_auto_enroll_course_ids = ( new LLMS_Membership( $membership_id ) )->get( 'auto_enroll' );
 
-			// Loop through all the courses and update the enrollment status.
-			foreach ( $courses  as $course_id ) {
-				if ( ! $delete ) {
-					$this->unenroll( $course_id, 'membership_' . $membership_id, $status );
-				} else {
-					$this->delete_enrollment( $course_id, 'membership_' . $membership_id );
-				}
+		// Loop through all the auto-enrolled courses and update the enrollment status.
+		foreach ( $given_membership_auto_enroll_course_ids as $given_membership_auto_enroll_course_id ) {
+
+			// Do not change course enrollment if the user does not have one. It may have been deleted manually.
+			if ( ! in_array( $given_membership_auto_enroll_course_id, $course_ids_enrolled_by_any_membership, true ) ) {
+				continue;
+			}
+
+			// Do not change course enrollment if the course was auto-enrolled by another existing membership.
```

Comment:
> Here we're testing that the course id (auto-enroll of this membership, ~and that triggered the enrollment~) is not in the auto-enroll list of another membership this user is enrolled into.

Actually, we're checking to see if another membership may have triggered the enrollment.

This check is the key to the whole issue. If the auto-enroll course for the given membership is in the array of auto-enroll courses for the other memberships that the user is enrolled in (status = enrolled), then leave the enrollment alone.
```suggestion
			// Do not change course enrollment if the user may have been auto-enrolled in the course by another membership
			// that still exists and that the user is still enrolled in (status = active).
```
I agree 100% that comments should help us. I struggled on this PR explaining things in a short and easy to read way. It's complicated.

> Basically we don't know whether or not the other membership's auto-enroll course could've triggered an enrollment: e.g. the course was added to the other membership's auto-enroll courses after the student was enrolled in the membership. This is a limit of our enrollments data structure.

¡Ay caramba! I hadn't thought about courses added to the auto-enroll list **after** the user is enrolled in the membership.

I'll work on this some more.

Reply:
> Actually, we're checking to see if another membership may have triggered the enrollment.

Not really, I mean, we're sure that `$given_membership_auto_enroll_course_id` at that point triggered the enrollment because the previous check excluded all this membership's auto-enroll course ids which are not enrollment_triggers:
```
if ( ! in_array( $given_membership_auto_enroll_course_id, $course_ids_enrolled_by_any_membership, true ) ) {
	continue;
}
```
Right?

> This check is the key to the whole issue. If the auto-enroll course for the given membership is in the array of auto-enroll courses for the other memberships that the user is enrolled in (status = enrolled), then leave the enrollment alone.

Yeah, I got it :P 

> ¡Ay caramba! I hadn't thought about courses added to the auto-enroll list after the user is enrolled in the membership.
> 
> I'll work on this some more.

Oh I'm not sure we can do that much on this.

Reply:
> > Actually, we're checking to see if another membership may have triggered the enrollment.
> 
> Not really, I mean, we're sure that `$given_membership_auto_enroll_course_id` at that point triggered the enrollment because the previous check excluded all this membership's auto-enroll course ids which are not enrollment_triggers:
> 
> ```
> if ( ! in_array( $given_membership_auto_enroll_course_id, $course_ids_enrolled_by_any_membership, true ) ) {
> 	continue;
> }
> ```
> 
> Right?

No. The first check excludes a given membership's auto-enroll course if the course enrollment was not triggered by any membership.

Suggested change to the first check's comment:
```
// Do not change course enrollment if the user does not have one. It may have been deleted manually.

// Do not change course enrollment if not triggered by a membership.
```



If the auto-enroll course for this membership is added to the auto-enroll courses of another membership after the user is enrolled in that membership,

Reply:
Even better comment for check #1?
`// Skip changes to course enrollments if the enrollment was not triggered by a membership.`